### PR TITLE
feat(lsp): add range option to lsp.buf.format

### PR DESCRIPTION
**Pull request by mfussenegger**
Alternative to https://github.com/neovim/neovim/pull/19736

### CODE REVIEW

**Comment by justinmk** [comment]
> Alternative to #19736

How is it an alternative? Isn't deprecation of formatting() still on the table?

**Comment by mfussenegger** [comment]
> > Alternative to #19736
> 
> How is it an alternative? Isn't deprecation of formatting() still on the table?

It is. Sorry should've been more verbose in the PR description.
This PR also deprecates range_formatting, but in addition it adds the range param for format as an alternative replacement in addition to formatexpr. To support the cases mentioned in https://github.com/neovim/neovim/pull/19736#issuecomment-1212886012

File: runtime/lua/vim/lsp/buf.lua [hunk 1]
```diff
@@ -150,6 +150,31 @@ local function select_client(method, on_choice)
   end
 end
 
+---@private
+---@return table {start={row, col}, end={row, col}} using (1, 0) indexing
+local function range_from_selection()
```
Comment (justinmk) [inline runtime/lua/vim/lsp/buf.lua:155]: Would you be interested in taking over https://github.com/neovim/neovim/pull/13896 ? Meanwhile should there be a comment here like:

    -- TODO: use `vim.region()` instead https://github.com/neovim/neovim/pull/13896

Comment (mfussenegger) [reply runtime/lua/vim/lsp/buf.lua:155]: Added the TODO. Currently I don't plan on taking over that PR, but that might change if I want this functionality in one of my plugins.

File: runtime/lua/vim/lsp/buf.lua [hunk 2]
```diff
@@ -206,16 +238,32 @@ function M.format(options)
     vim.notify('[LSP] Format request failed, no matching language servers.')
   end
 
+  local mode = api.nvim_get_mode().mode
```
[the remark this answers is not in the document]
Comment (GZLiew) [reply runtime/lua/vim/lsp/buf.lua:241]: for some reason this doesnt work for me
mode is always `'n'` and proceeds to format the whole file for me
my neovim is at 0.8.0

[the remark this answers is not in the document]
Comment (mfussenegger) [reply runtime/lua/vim/lsp/buf.lua:241]: @GZLiew you're probably using a mapping including `<cmd>` or `<esc>` or something like that. You need to stay in visual mode for the mapping to work.

For example:

```lua
vim.keymap.set('v', 'gq', vim.lsp.buf.format)
```

**Pull request merged**